Added DSLClock

Vader-Femi opened the pull request:
- Added DSL clock faces for time and complications. 
- Added permission for BODY_SENSORS in app-watch manifest file

Vader-Femi committed:
Added DCLClock

Added DSL clock faces for time and complications

yschimke commented on shared/src/main/kotlin/watchfaces/dslclock/DSLClockComplication.kt:
```diff
@@ -0,0 +1,355 @@
+    var hr by remember { mutableIntStateOf(0) }
+
+    val sensorManager = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
+    val hrCounter = sensorManager.getDefaultSensor(Sensor.TYPE_HEART_RATE)
```
It seems a bad idea to be using SensorManager in a Watchface.

Could you use either of
- Complications
- WHS PassiveClient ?

Vader-Femi replied:
Okay. I will look into that

LouisCAD replied:
FYI, "WHS" stands for "Wear Health Services", and the doc on it and passive client is here: https://developer.android.com/health-and-fitness/guides/health-services/compatibility#passive_monitoring_client

Vader-Femi replied:
I used WHS initially, but it required me to increase the minSdk to 30, so I removed it

LouisCAD commented on shared/src/main/kotlin/watchfaces/dslclock/DSLClockComplication.kt:
```diff
@@ -0,0 +1,355 @@
+@Composable
+private fun Battery(style: TextStyle) {
+
+    val context: Context = LocalContext.current
+    var batteryLevel by remember { mutableIntStateOf(0) }
+
+    val iFilter = IntentFilter(Intent.ACTION_BATTERY_CHANGED)
+    val batteryStatus = context.registerReceiver(null, iFilter)
+
+    batteryLevel = batteryStatus?.getIntExtra(BatteryManager.EXTRA_LEVEL, -1) ?: -1
+
+    context.registerReceiver(
+        object : BroadcastReceiver() {
+            override fun onReceive(ctxt: Context, intent: Intent) {
+                try {
+                    batteryLevel = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1)
+                } catch (_: IllegalStateException) {
+                }
+            }
+        },
+        IntentFilter(Intent.ACTION_BATTERY_CHANGED))
```
You wouldn't want to run this code on every recomposition.
Most of that needs to be put inside a `LaunchedEffect`, a `DisposableEffect`, or a `produceState`, as usual in Compose.

LouisCAD commented on shared/src/main/kotlin/watchfaces/dslclock/DSLClockComplication.kt:
```diff
@@ -0,0 +1,355 @@
+    val isAmbient by LocalIsAmbient.current
+    var hr by remember { mutableIntStateOf(0) }
+
+    val sensorManager = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
+    val hrCounter = sensorManager.getDefaultSensor(Sensor.TYPE_HEART_RATE)
```
Same as above in `Battery`, but we probably want to replace that with use of Complications API or WHS PassiveClient anyway.

LouisCAD commented on shared/src/main/kotlin/watchfaces/dslclock/DSLClockComplication.kt:
```diff
@@ -0,0 +1,355 @@
+    }
+
+    OClockCanvas {
+        rotate(degrees = 0f) {
```
Why rotate if the angle ends up being zero all the time?

Vader-Femi replied:
Fair enough

LouisCAD commented on shared/src/main/kotlin/watchfaces/dslclock/DSLClockComplication.kt:
```diff
@@ -0,0 +1,355 @@
+                brush = style.brush ?: SolidColor(style.color),
+                textLayoutResult = textLayoutResult,
+                topLeft = Offset(
+                    x = 120f,
```
Hardcoded pixel size that is not dependent on the canvas size is gonna be problematic since not all screens have the same density (and resolution).

LouisCAD commented on shared/src/main/kotlin/watchfaces/dslclock/DSLClockComplication.kt:
```diff
@@ -0,0 +1,355 @@
+                brush = style.brush ?: SolidColor(style.color),
+                textLayoutResult = textLayoutResult,
+                topLeft = Offset(
+                    x = 120f,
```
Same as above

LouisCAD commented on shared/src/main/kotlin/watchfaces/dslclock/DSLClockComplication.kt:
```diff
@@ -0,0 +1,355 @@
+                brush = style.brush ?: SolidColor(style.color),
+                textLayoutResult = textLayoutResult,
+                topLeft = Offset(
+                    x = 60f,
```
Same as above

LouisCAD commented on shared/src/main/kotlin/watchfaces/dslclock/DSLClockTime.kt:
```diff
@@ -0,0 +1,195 @@
+            drawText(
+                brush = style.brush ?: SolidColor(style.color),
+                textLayoutResult = textLayoutResult,
+                topLeft = Offset(x = 60f, y = (center.y * 0.55f) - (textLayoutResult.size.height / 2f))
```
Same as above

LouisCAD commented on shared/src/main/kotlin/watchfaces/dslclock/DSLClockTime.kt:
```diff
@@ -0,0 +1,195 @@
+            drawText(
+                brush = style.brush ?: SolidColor(style.color),
+                textLayoutResult = textLayoutResult,
+                topLeft = Offset(x = 60f, y = size.height - (center.y * 0.55f)  - (textLayoutResult.size.height / 2f))
```
Same as above

LouisCAD commented on shared/src/main/kotlin/watchfaces/dslclock/DSLClockTime.kt:
```diff
@@ -0,0 +1,195 @@
+    OClockCanvas {
+        rotate(degrees = 0f) {
+            val offset = when (timeType) {
+                TimeParamType.Hour -> Offset(x = 140f, y = (center.y * 0.8f) - (textLayoutResult.size.height / 2f))
+                TimeParamType.Minute -> Offset(x = 140f, y = center.y - (textLayoutResult.size.height / 2f))
+                TimeParamType.Second -> Offset(x = 140f, y = size.height - (center.y * 0.8f) - (textLayoutResult.size.height / 2f))
```
Same as above

Vader-Femi committed:
Removed SensorEventListener from heart rate data

Vader-Femi committed:
Removed heart rate and bug fixes

- Removed Heart rate from complications face

- Removed unused rotation from various OClockCanvas

- Changed date format to "yyyy/MM/dd" for shorter length

- Removed hardcoded offset values

- Kept battery BroadcastReceiver in a DisposableEffect
